cloud/cloud.go

```diff
@@ -113,8 +113,13 @@ type ContainerManager interface {
 
 // BatchManager is an interface for cloud providers that support batch operations.
 type BatchManager interface {
-	// GetInstanceStatuses gets the status of a slice of instances.
-	GetInstanceStatuses(context.Context, []host.Host) ([]CloudStatus, error)
+	// GetInstanceStatuses gets the statuses of a slice of instances. It returns
+	// a map of the instance IDs to their current status. If some of the
+	// instance statuses cannot be retrieved, implementations are allowed to
+	// either return an error or return StatusNonExistent for those hosts.
+	// If there is no error, implementations should return the same number of
+	// results in the map as there are hosts.
```

Comment:
Why is it desirable for implementations to have this flexibility?

Reply:
This is just how the AWS response behaves. In the past, it has returned an error if any of the instance IDs does not exist, but in the case of this outage, it returned no error but failed to return some statuses.

In the end, past experience has proven that it's unreliable to trust that AWS will return us what we think will be their response, so might as well make it as resilient as possible to strange behavior.

Reply:
ah I get it now.
Could it make sense for `GetInstanceStatuses` to never error for non-existent hosts and instead always return the non-existent hosts with the `StatusNonExistent` status? It anyways seems to me that the AWS error string parsing we do to get the offending instance ids ([this thing](https://github.com/evergreen-ci/evergreen/blob/f486e5cef70c4425dedb6b974d35597d13f14341/units/host_status.go#L147-L151)) belongs somewhere in ec2.go rather than host_status.go. 
It would have an additional advantage that we could consolidate `terminateUnknownHosts` into `setCloudHostStatus`.

Reply:
We could, but it would make the logic more complicated we'll have to push the error-parsing logic into `GetInstanceStatuses` and re-issue the `DescribeInstances` request with the smaller set of instance IDs (and repeatedly do it if we keep getting nonexistent hosts). I think it would be nice to have, but it seems more like a small improvement of the abstraction layers that I'd rather leave for a later time. My primary goal with this PR is to ensure that `GetInstanceStatuses` is more resilient this new response.

Reply:
https://jira.mongodb.org/browse/EVG-18280